lib/api/core/janitor.js

```diff
@@ -0,0 +1,368 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2017 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable no-console */
+
+'use strict';
+
+const
+  Bluebird = require('bluebird'),
+  Request = require('kuzzle-common-objects').Request,
+  {
+    BadRequestError,
+    InternalError: KuzzleInternalError,
+  } = require('kuzzle-common-objects').errors,
+  pm2 = require('pm2'),
+  path = require('path'),
+  fs = require('fs-extra'),
+  os = require('os'),
+  moment = require('moment'),
+  dumpme = require('dumpme'),
+  zlib = require('zlib'),
+  glob = require('glob');
+
+
+class Janitor {
+  constructor (kuzzle) {
+    this.kuzzle = kuzzle;
+
+    this._shutdown = 0;
+    this._dump = false;
+  }
+
+  /**
+   * Gracefully exits after processing remaining requests
+   *
+   * @param {Kuzzle} kuzzle
+   * @returns {Promise}
+   */
+  shutdown () {
+    // pm2.delete sends a SIGINT signal
+    // we keep track of how many shutdowns are asked
+    // to detect when pm2 has finished deleting Kuzzle
+    // from its tasks list
+    this._shutdown++;
+
+    if (this._shutdown > 1) {
+      return Bluebird.resolve();
+    }
+
+    console.log('Initiating shutdown...');
+
+    return new Bluebird(resolve => {
+      this.kuzzle.entryPoints.dispatch('shutdown');
+
+      // Gives time for the proxy to receive the shutdown signal
+      // and eventually to receive the last batch of requests to process
+      setTimeout(() => this._waitRemainingRequests(resolve), 1000);
+    });
+  }
+
+  /**
+   * Create a dump
+   *
+   * @param {string} suffix
+   * @returns {Promise}
+   */
+  dump (suffix) {
+    if (this._dump) {
+      throw new BadRequestError('A dump is already being generated. Skipping.');
+    }
+
+    if (! this.kuzzle.config.dump.enabled) {
+      throw new BadRequestError('Dump generation is disabled.');
+    }
+
+    let dumpPath;
+    this._dump = true;
+    suffix = `-${suffix}`;
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Introduce a new variable instead of reusing the parameter "suffix". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1226)